File: lua/dingconnect_sdk.lua
-- Dingconnect SDK

local vs = require("utility.struct.struct")
local Utility = require("core.utility_type")
local Spec = require("core.spec")
local helpers = require("core.helpers")

-- Load utility registration (populates Utility._registrar)
require("utility.register")

-- Typed-model annotations (LuaLS ---@class); empty at runtime.
require("dingconnect_types")

-- Load features
local BaseFeature = require("feature.base_feature")
local features_factory = require("features")


local DingconnectSDK = {}
DingconnectSDK.__index = DingconnectSDK


local function _make_feature(name)
  local factory = features_factory[name]
  if factory ~= nil then
    return factory()
  end
  return features_factory.base()
end

DingconnectSDK._make_feature = _make_feature


function DingconnectSDK.new(options)
  local self = setmetatable({}, DingconnectSDK)
  self.mode = "live"
  self.features = {}
  self.options = nil

  local utility = Utility.new()
  self._utility = utility

  local config = require("config")()

  self._rootctx = utility.make_context({
    client = self,
    utility = utility,
    config = config,
    options = options or {},
    shared = {},
  }, nil)

  self.options = utility.make_options(self._rootctx)

  if vs.getpath(self.options, "feature.test.active") == true then
    self.mode = "test"
  end

  self._rootctx.options = self.options

  -- Add features in the resolved order (make_options puts an explicit list
  -- order first, else defaults to test-first). Ordering matters: the `test`
  -- feature installs the base mock transport and the transport features
  -- (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
  -- must be added before them to sit at the base of the chain.
  local feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
  if feature_opts ~= nil then
    local featureorder = vs.getpath(self.options, "__derived__.featureorder")
    if type(featureorder) == "table" then
      for _, fname in ipairs(featureorder) do
        local fopts = helpers.to_map(feature_opts[fname])
        if fopts ~= nil and fopts["active"] == true then
          utility.feature_add(self._rootctx, _make_feature(fname))
        end
      end
    end
  end

  -- Add extension features.
  local extend = vs.getprop(self.options, "extend")
  if type(extend) == "table" then
    for _, f in ipairs(extend) do
      if type(f) == "table" and type(f.get_name) == "function" then
        utility.feature_add(self._rootctx, f)
      end
    end
  end

  -- Initialize features.
  for _, f in ipairs(self.features) do
    utility.feature_init(self._rootctx, f)
  end

  utility.feature_hook(self._rootctx, "PostConstruct")

    -- feature: test


  return self
end


function DingconnectSDK:options_map()
  local out = vs.clone(self.options)
  if type(out) == "table" then
    return out
  end
  return {}
end


function DingconnectSDK:get_utility()
  return Utility.copy(self._utility)
end


function DingconnectSDK:get_root_ctx()
  return self._rootctx
end


function DingconnectSDK:prepare(fetchargs)
  local utility = self._utility

  fetchargs = fetchargs or {}

  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "prepare",
    ctrl = ctrl,
  }, self._rootctx)

  local options = self.options

  local path = vs.getprop(fetchargs, "path") or ""
  if type(path) ~= "string" then path = "" end

  local method = vs.getprop(fetchargs, "method") or "GET"
  if type(method) ~= "string" then method = "GET" end

  local params = helpers.to_map(vs.getprop(fetchargs, "params")) or {}
  local query = helpers.to_map(vs.getprop(fetchargs, "query")) or {}

  local headers = utility.prepare_headers(ctx)

  local base = vs.getprop(options, "base") or ""
  if type(base) ~= "string" then base = "" end
  local prefix = vs.getprop(options, "prefix") or ""
  if type(prefix) ~= "string" then prefix = "" end
  local suffix = vs.getprop(options, "suffix") or ""
  if type(suffix) ~= "string" then suffix = "" end

  ctx.spec = Spec.new({
    base = base,
    prefix = prefix,
    suffix = suffix,
    path = path,
    method = method,
    params = params,
    query = query,
    headers = headers,
    body = vs.getprop(fetchargs, "body"),
    step = "start",
  })

  -- Merge user-provided headers.
  local uh = vs.getprop(fetchargs, "headers")
  if type(uh) == "table" then
    for k, v in pairs(uh) do
      ctx.spec.headers[k] = v
    end
  end

  local _, err = utility.prepare_auth(ctx)
  if err ~= nil then
    return nil, err
  end

  return utility.make_fetch_def(ctx)
end


function DingconnectSDK:direct(fetchargs)
  local utility = self._utility

  local fetchdef, err = self:prepare(fetchargs)
  if err ~= nil then
    return { ok = false, err = err }, nil
  end

  fetchargs = fetchargs or {}
  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "direct",
    ctrl = ctrl,
  }, self._rootctx)

  local url = fetchdef["url"] or ""
  local fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

  if fetch_err ~= nil then
    return { ok = false, err = fetch_err }, nil
  end

  if fetched == nil then
    return {
      ok = false,
      err = ctx:make_error("direct_no_response", "response: undefined"),
    }, nil
  end

  if type(fetched) == "table" then
    local status = helpers.to_int(vs.getprop(fetched, "status"))
    local headers = vs.getprop(fetched, "headers") or {}

    -- No-body responses (204, 304) and explicit zero content-length
    -- must skip JSON parsing — calling json() on an empty body errors.
    local content_length = nil
    if type(headers) == "table" then
      content_length = headers["content-length"]
    end
    local no_body = status == 204 or status == 304 or tostring(content_length) == "0"

    local json_data = nil
    if not no_body then
      local jf = vs.getprop(fetched, "json")
      if type(jf) == "function" then
        local ok, result = pcall(jf)
        if ok then
          json_data = result
        end
        -- Non-JSON body: json_data stays nil, status/headers preserved.
      end
    end

    return {
      ok = status >= 200 and status < 300,
      status = status,
      headers = headers,
      data = json_data,
    }, nil
  end

  return {
    ok = false,
    err = ctx:make_error("direct_invalid", "invalid response type"),
  }, nil
end



-- Idiomatic facade: client:AccountLookup():list() / client:AccountLookup():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function DingconnectSDK:AccountLookup(data)
  local EntityMod = require("entity.account_lookup_entity")
  if data == nil then
    if self._account_lookup == nil then
      self._account_lookup = EntityMod.new(self, nil)
    end
    return self._account_lookup
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Balance():list() / client:Balance():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function DingconnectSDK:Balance(data)
  local EntityMod = require("entity.balance_entity")
  if data == nil then
    if self._balance == nil then
      self._balance = EntityMod.new(self, nil)
    end
    return self._balance
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:CancelResult():list() / client:CancelResult():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function DingconnectSDK:CancelResult(data)
  local EntityMod = require("entity.cancel_result_entity")
  if data == nil then
    if self._cancel_result == nil then
      self._cancel_result = EntityMod.new(self, nil)
    end
    return self._cancel_result
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Country():list() / client:Country():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function DingconnectSDK:Country(data)
  local EntityMod = require("entity.country_entity")
  if data == nil then
    if self._country == nil then
      self._country = EntityMod.new(self, nil)
    end
    return self._country
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Currency():list() / client:Currency():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function DingconnectSDK:Currency(data)
  local EntityMod = require("entity.currency_entity")
  if data == nil then
    if self._currency == nil then
      self._currency = EntityMod.new(self, nil)
    end
    return self._currency
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:ErrorCodeDescription():list() / client:ErrorCodeDescription():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function DingconnectSDK:ErrorCodeDescription(data)
  local EntityMod = require("entity.error_code_description_entity")
  if data == nil then
    if self._error_code_description == nil then
      self._error_code_description = EntityMod.new(self, nil)
    end
    return self._error_code_description
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Estimate():list() / client:Estimate():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function DingconnectSDK:Estimate(data)
  local EntityMod = require("entity.estimate_entity")
  if data == nil then
    if self._estimate == nil then
      self._estimate = EntityMod.new(self, nil)
    end
    return self._estimate
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Product():list() / client:Product():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function DingconnectSDK:Product(data)
  local EntityMod = require("entity.product_entity")
  if data == nil then
    if self._product == nil then
      self._product = EntityMod.new(self, nil)
    end
    return self._product
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:ProductDescription():list() / client:ProductDescription():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function DingconnectSDK:ProductDescription(data)
  local EntityMod = require("entity.product_description_entity")
  if data == nil then
    if self._product_description == nil then
      self._product_description = EntityMod.new(self, nil)
    end
    return self._product_description
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Promotion():list() / client:Promotion():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function DingconnectSDK:Promotion(data)
  local EntityMod = require("entity.promotion_entity")
  if data == nil then
    if self._promotion == nil then
      self._promotion = EntityMod.new(self, nil)
    end
    return self._promotion
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:PromotionDescription():list() / client:PromotionDescription():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function DingconnectSDK:PromotionDescription(data)
  local EntityMod = require("entity.promotion_description_entity")
  if data == nil then
    if self._promotion_description == nil then
      self._promotion_description = EntityMod.new(self, nil)
    end
    return self._promotion_description
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Provider():list() / client:Provider():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function DingconnectSDK:Provider(data)
  local EntityMod = require("entity.provider_entity")
  if data == nil then
    if self._provider == nil then
      self._provider = EntityMod.new(self, nil)
    end
    return self._provider
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:ProviderStatus():list() / client:ProviderStatus():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function DingconnectSDK:ProviderStatus(data)
  local EntityMod = require("entity.provider_status_entity")
  if data == nil then
    if self._provider_status == nil then
      self._provider_status = EntityMod.new(self, nil)
    end
    return self._provider_status
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Region():list() / client:Region():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function DingconnectSDK:Region(data)
  local EntityMod = require("entity.region_entity")
  if data == nil then
    if self._region == nil then
      self._region = EntityMod.new(self, nil)
    end
    return self._region
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:SendTransfer():list() / client:SendTransfer():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function DingconnectSDK:SendTransfer(data)
  local EntityMod = require("entity.send_transfer_entity")
  if data == nil then
    if self._send_transfer == nil then
      self._send_transfer = EntityMod.new(self, nil)
    end
    return self._send_transfer
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:TransferRecord():list() / client:TransferRecord():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function DingconnectSDK:TransferRecord(data)
  local EntityMod = require("entity.transfer_record_entity")
  if data == nil then
    if self._transfer_record == nil then
      self._transfer_record = EntityMod.new(self, nil)
    end
    return self._transfer_record
  end
  return EntityMod.new(self, data)
end




function DingconnectSDK.test(testopts, sdkopts)
  sdkopts = sdkopts or {}
  sdkopts = vs.clone(sdkopts)
  if type(sdkopts) ~= "table" then
    sdkopts = {}
  end

  testopts = testopts or {}
  testopts = vs.clone(testopts)
  if type(testopts) ~= "table" then
    testopts = {}
  end
  testopts["active"] = true

  vs.setpath(sdkopts, "feature.test", testopts)

  local sdk = DingconnectSDK.new(sdkopts)
  sdk.mode = "test"

  return sdk
end


return DingconnectSDK

File: ts/src/DingconnectSDK.ts
// Dingconnect Ts SDK

import { AccountLookupEntity } from './entity/AccountLookupEntity'
import { BalanceEntity } from './entity/BalanceEntity'
import { CancelResultEntity } from './entity/CancelResultEntity'
import { CountryEntity } from './entity/CountryEntity'
import { CurrencyEntity } from './entity/CurrencyEntity'
import { ErrorCodeDescriptionEntity } from './entity/ErrorCodeDescriptionEntity'
import { EstimateEntity } from './entity/EstimateEntity'
import { ProductEntity } from './entity/ProductEntity'
import { ProductDescriptionEntity } from './entity/ProductDescriptionEntity'
import { PromotionEntity } from './entity/PromotionEntity'
import { PromotionDescriptionEntity } from './entity/PromotionDescriptionEntity'
import { ProviderEntity } from './entity/ProviderEntity'
import { ProviderStatusEntity } from './entity/ProviderStatusEntity'
import { RegionEntity } from './entity/RegionEntity'
import { SendTransferEntity } from './entity/SendTransferEntity'
import { TransferRecordEntity } from './entity/TransferRecordEntity'

export type * from './DingconnectTypes'


import { inspect } from 'node:util'

import type { Context, Feature } from './types'

import { config } from './Config'
import { DingconnectEntityBase } from './DingconnectEntityBase'
import { Utility } from './utility/Utility'


import { BaseFeature } from './feature/base/BaseFeature'


const stdutil = new Utility()


class DingconnectSDK {
  _mode: string = 'live'
  _options: any
  _utility = new Utility()
  _features: Feature[]
  _rootctx: Context

  constructor(options?: any) {

    this._rootctx = this._utility.makeContext({
      client: this,
      utility: this._utility,
      config,
      options,
      shared: new WeakMap()
    })

    this._options = this._utility.makeOptions(this._rootctx)

    const struct = this._utility.struct
    const getpath = struct.getpath

    if (true === getpath(this._options.feature, 'test.active')) {
      this._mode = 'test'
    }

    this._rootctx.options = this._options

    this._features = []

    const featureAdd = this._utility.featureAdd
    const featureInit = this._utility.featureInit

    // Add features in the resolved order (makeOptions puts an explicit
    // array order first, else defaults to test-first). Ordering matters:
    // the `test` feature installs the base mock transport and the transport
    // features (retry/cache/netsim/proxy/ratelimit) wrap whatever is current,
    // so `test` must be added before them to sit at the base of the chain.
    const featureorder = getpath(this._options, '__derived__.featureorder') || []
    for (const fname of featureorder) {
      const fopts = this._options.feature[fname] || {}
      if (fopts.active) {
        featureAdd(this._rootctx, this._rootctx.config.makeFeature(fname))
      }
    }

    if (null != this._options.extend) {
      for (let f of this._options.extend) {
        featureAdd(this._rootctx, f)
      }
    }

    for (let f of this._features) {
      featureInit(this._rootctx, f)
    }

    const featureHook = this._utility.featureHook
    featureHook(this._rootctx, 'PostConstruct')
  }


  options() {
    return this._utility.struct.clone(this._options)
  }


  utility() {
    return this._utility.struct.clone(this._utility)
  }


  async prepare(fetchargs?: any) {
    const utility = this._utility
    const struct = utility.struct
    const clone = struct.clone

    const {
      makeContext,
      makeFetchDef,
      prepareHeaders,
      prepareAuth,
    } = utility

    fetchargs = fetchargs || {}

    let ctx: Context = makeContext({
      opname: 'prepare',
      ctrl: fetchargs.ctrl || {},
    }, this._rootctx)

    const options = this._options

    // Build spec directly from SDK options + user-provided fetch args.
    const spec: any = {
      base: options.base,
      prefix: options.prefix,
      suffix: options.suffix,
      path: fetchargs.path || '',
      method: fetchargs.method || 'GET',
      params: fetchargs.params || {},
      query: fetchargs.query || {},
      headers: prepareHeaders(ctx),
      body: fetchargs.body,
      step: 'start',
    }

    ctx.spec = spec

    // Merge user-provided headers over SDK defaults.
    if (fetchargs.headers) {
      const uheaders = fetchargs.headers
      for (let key in uheaders) {
        spec.headers[key] = uheaders[key]
      }
    }

    // Apply SDK auth (apikey, auth prefix, etc.)
    const authResult = prepareAuth(ctx)
    if (authResult instanceof Error) {
      return authResult
    }

    return makeFetchDef(ctx)
  }


  async direct(fetchargs?: any) {
    const utility = this._utility
    const fetcher = utility.fetcher
    const makeContext = utility.makeContext

    const fetchdef = await this.prepare(fetchargs)
    if (fetchdef instanceof Error) {
      return fetchdef
    }

    let ctx: Context = makeContext({
      opname: 'direct',
      ctrl: (fetchargs || {}).ctrl || {},
    }, this._rootctx)

    try {
      const fetched = await fetcher(ctx, fetchdef.url, fetchdef)

      if (null == fetched) {
        return { ok: false, err: ctx.error('direct_no_response', 'response: undefined') }
      }
      else if (fetched instanceof Error) {
        return { ok: false, err: fetched }
      }

      const status = fetched.status

      // No body responses (204 No Content, 304 Not Modified) and explicit
      // zero content-length must skip JSON parsing — fetched.json() would
      // throw `Unexpected end of JSON input` on an empty body.
      const headers = fetched.headers
      const contentLength = headers && 'function' === typeof headers.get
        ? headers.get('content-length')
        : (headers || {})['content-length']
      const noBody = 204 === status || 304 === status || '0' === String(contentLength)

      let json: any = undefined
      if (!noBody) {
        try {
          json = 'function' === typeof fetched.json ? await fetched.json() : fetched.json
        }
        catch (parseErr) {
          // Body wasn't valid JSON — surface the raw response rather than
          // throwing. data stays undefined; callers can inspect status/headers.
          json = undefined
        }
      }

      return {
        ok: status >= 200 && status < 300,
        status,
        headers: fetched.headers,
        data: json,
      }
    }
    catch (err: any) {
      return { ok: false, err }
    }
  }



  // Entity access: `client.AccountLookup().list()` / `client.AccountLookup().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  AccountLookup(entopts?: Record<string, any>) {
    const self = this
    return new AccountLookupEntity(self, entopts)
  }


  // Entity access: `client.Balance().list()` / `client.Balance().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Balance(entopts?: Record<string, any>) {
    const self = this
    return new BalanceEntity(self, entopts)
  }


  // Entity access: `client.CancelResult().list()` / `client.CancelResult().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  CancelResult(entopts?: Record<string, any>) {
    const self = this
    return new CancelResultEntity(self, entopts)
  }


  // Entity access: `client.Country().list()` / `client.Country().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Country(entopts?: Record<string, any>) {
    const self = this
    return new CountryEntity(self, entopts)
  }


  // Entity access: `client.Currency().list()` / `client.Currency().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Currency(entopts?: Record<string, any>) {
    const self = this
    return new CurrencyEntity(self, entopts)
  }


  // Entity access: `client.ErrorCodeDescription().list()` / `client.ErrorCodeDescription().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  ErrorCodeDescription(entopts?: Record<string, any>) {
    const self = this
    return new ErrorCodeDescriptionEntity(self, entopts)
  }


  // Entity access: `client.Estimate().list()` / `client.Estimate().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Estimate(entopts?: Record<string, any>) {
    const self = this
    return new EstimateEntity(self, entopts)
  }


  // Entity access: `client.Product().list()` / `client.Product().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Product(entopts?: Record<string, any>) {
    const self = this
    return new ProductEntity(self, entopts)
  }


  // Entity access: `client.ProductDescription().list()` / `client.ProductDescription().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  ProductDescription(entopts?: Record<string, any>) {
    const self = this
    return new ProductDescriptionEntity(self, entopts)
  }


  // Entity access: `client.Promotion().list()` / `client.Promotion().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Promotion(entopts?: Record<string, any>) {
    const self = this
    return new PromotionEntity(self, entopts)
  }


  // Entity access: `client.PromotionDescription().list()` / `client.PromotionDescription().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  PromotionDescription(entopts?: Record<string, any>) {
    const self = this
    return new PromotionDescriptionEntity(self, entopts)
  }


  // Entity access: `client.Provider().list()` / `client.Provider().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Provider(entopts?: Record<string, any>) {
    const self = this
    return new ProviderEntity(self, entopts)
  }


  // Entity access: `client.ProviderStatus().list()` / `client.ProviderStatus().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  ProviderStatus(entopts?: Record<string, any>) {
    const self = this
    return new ProviderStatusEntity(self, entopts)
  }


  // Entity access: `client.Region().list()` / `client.Region().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Region(entopts?: Record<string, any>) {
    const self = this
    return new RegionEntity(self, entopts)
  }


  // Entity access: `client.SendTransfer().list()` / `client.SendTransfer().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  SendTransfer(entopts?: Record<string, any>) {
    const self = this
    return new SendTransferEntity(self, entopts)
  }


  // Entity access: `client.TransferRecord().list()` / `client.TransferRecord().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  TransferRecord(entopts?: Record<string, any>) {
    const self = this
    return new TransferRecordEntity(self, entopts)
  }




  static test(testoptsarg?: any, sdkoptsarg?: any) {
    const struct = stdutil.struct
    const setpath = struct.setpath
    const getdef = struct.getdef
    const clone = struct.clone
    const setprop = struct.setprop

    const sdkopts = getdef(clone(sdkoptsarg), {})
    const testopts = getdef(clone(testoptsarg), {})
    setprop(testopts, 'active', true)
    setpath(sdkopts, 'feature.test', testopts)

    const testsdk = new DingconnectSDK(sdkopts)
    testsdk._mode = 'test'

    return testsdk
  }


  tester(testopts?: any, sdkopts?: any) {
    return DingconnectSDK.test(testopts, sdkopts)
  }


  toJSON() {
    return { name: 'Dingconnect' }
  }

  toString() {
    return 'Dingconnect ' + this._utility.struct.jsonify(this.toJSON())
  }

  [inspect.custom]() {
    return this.toString()
  }

}




const SDK = DingconnectSDK


export {
  stdutil,
  config,

  BaseFeature,
  DingconnectEntityBase,

  DingconnectSDK,
  SDK,
}


